src/pipeline-stages/src/main/scala/TextPreprocessor.scala

```diff
@@ -0,0 +1,141 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.param.{MapParam, Param, ParamMap}
+import org.apache.spark.ml.util.{ComplexParamsReadable, ComplexParamsWritable, Identifiable}
+import org.apache.spark.sql.types.{StringType, StructField, StructType}
+import org.apache.spark.sql.{DataFrame, Dataset}
+import org.apache.spark.sql.functions.udf
+import spray.json.DefaultJsonProtocol._
+
+class Trie(map: Map[Char, Trie] = Map.empty,
+           textValue: Array[Char] = Array.empty,
+           normFunction: Char => Char = identity) extends Serializable {
+  val trieMap: Map[Char, Trie] = map
+  val value: Array[Char] = textValue
+  val normFunc: Char => Char = normFunction
+  def getValueString: String = this.value.toString
+
+  def getNodeValue(key: Array[Char], value: Array[Char]): Array[Char] = (key, value) match {
```

Comment:
Please consider using ``Seq`` instead of ``Array``

Reply:
I made the fix. Only needed length like information for one location, i used empty tail as a idiomatic alternative to length == 1